fv3core/decorators.py

```diff
@@ -90,120 +92,116 @@ def wrapped(state, *args, **kwargs):
-                    **stencil_kwargs.get("externals", dict()),
-                }
-
-                axis_offsets = spec.grid.axis_offsets(
-                    origin=get_origin(func, args, kwargs)
-                )
-                stencil_kwargs["externals"].update(axis_offsets)
-
-                # Generate stencil
-                build_info = {}
-                stencil = gtscript.stencil(build_info=build_info, **stencil_kwargs)(
-                    func
-                )
-                stencils[key] = FV3StencilObject(stencil, build_info)
-            name = f"{func.__module__}.{func.__name__}"
-            _maybe_save_report(
-                f"{name}-before",
-                times_called,
-                func.__dict__["_gtscript_"]["api_signature"],
-                args,
-                kwargs,
-            )
-            kwargs["validate_args"] = kwargs.get("validate_args", utils.validate_args)
-            result = stencils[key](*args, **kwargs)
-            _maybe_save_report(
-                f"{name}-after",
-                times_called,
-                func.__dict__["_gtscript_"]["api_signature"],
-                args,
-                kwargs,
-            )
-            times_called += 1
-            return result
+    @property
+    def impl_ir(self) -> Optional[dict]:
+        """Current stencil implementation IR if built, else None."""
+        return self.build_info.impl_ir
+
+    def __call__(self, *args, origin: Int3, domain: Int3, **kwargs):
+        """Call the stencil, compiling the stencil if necessary.
+
+        The stencil needs to be recompiled if any of the following changes
+        1. the origin and/or domain
+        2. any external value
+        3. the function signature or code
+
+        Args:
+            domain: Stencil compute domain (required)
+            origin: Data index mapped to (0, 0, 0) in the compute domain (required)
+            externals: Dictionary of externals for the stencil call
+        """
+        axis_offsets = fv3core.utils.axis_offsets(spec.grid, origin, domain)
+
+        stencil_kwargs = {
+            "rebuild": global_config.get_rebuild()
+            or self.compute_domain != (origin, domain),
+            "backend": global_config.get_backend(),
+            "externals": {
+                "namelist": spec.namelist,
+                "grid": spec.grid,
+                **axis_offsets,
+                **self.externals,
+            },
+            **self.backend_kwargs,
+        }
 
-        return wrapped
+        # Generate stencil if out of date
+        new_build_info = {}
+        stencil_object = gtscript.stencil(
+            definition=self.func, build_info=new_build_info, **stencil_kwargs
+        )
+        if stencil_object != self.stencil_object:
+            self.build_info = new_build_info
+            self.stencil_object = stencil_object
+
+        # Call it
+        kwargs["validate_args"] = kwargs.get("validate_args", utils.validate_args)
```

Comment:
There's still room to optimize this in a later PR after more long-term design of compile time argument management. For example, if we store these in a container that keeps a boolean indicating whether the values have changed since last compilation or not, which gets updated when the values change, we can avoid re-creating these objects every call.